lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -843,6 +878,98 @@ void FunctionEntrypointAnalyzer::interproceduralPropagation() {
   }
 }
 
+std::optional<abi::RegisterState::Values>
+FEA::tryGetRegisterState(model::Register::Values RegisterValue,
+                         const ABIAnalyses::RegisterStateMap &ABIRegisterMap) {
+  using State = abi::RegisterState::Values;
+
+  llvm::StringRef Name = model::Register::getCSVName(RegisterValue);
+  if (llvm::GlobalVariable *CSV = M.getGlobalVariable(Name, true)) {
+    auto It = ABIRegisterMap.find(CSV);
+    if (It != ABIRegisterMap.end()) {
+      revng_assert(It->second != State::Count && It->second != State::Invalid);
+      return It->second;
+    }
+  }
+
+  return std::nullopt;
+}
+
+void FEA::initializeMapForDeductions(FunctionSummary &Summary,
+                                     abi::RegisterState::Map &Map) {
+  auto Arch = model::ABI::getArchitecture(Binary->DefaultABI);
+  revng_assert(Arch == Binary->Architecture);
+
+  for (const auto &Reg : model::Architecture::registers(Arch)) {
+    const auto &ArgRegisters = Summary.ABIResults.ArgumentsRegisters;
+    const auto &RVRegisters = Summary.ABIResults.FinalReturnValuesRegisters;
+
+    if (auto MaybeState = tryGetRegisterState(Reg, ArgRegisters))
+      Map[Reg].IsUsedForPassingArguments = *MaybeState;
+
+    if (auto MaybeState = tryGetRegisterState(Reg, RVRegisters))
+      Map[Reg].IsUsedForReturningValues = *MaybeState;
+  }
+}
+
+void FunctionEntrypointAnalyzer::applyABIDeductions() {
+  using namespace abi;
+  using State = RegisterState::Values;
+
+  if (ABIEnforcement.getNumOccurrences() == 0
+      || ABIEnforcement == NoABIEnforcement)
+    return;
+
+  for (const model::Function &Function : Binary->Functions) {
+    auto &Summary = Oracle.at(Function.Entry);
```

Comment:
* Did we already merge call-site and callee information at this stage?
* We should perform this kind of deductions for indirect function calls too.

Reply:
- We did, `interproceduralPropagation` has already run by the time `applyABIDeductions` runs.
- Uhmm, we are actually only doing this per function, not per call-site (the results of the ABI analysis per call-site are left untouched). Should we do enforce the ABI also for each call-site?